README.md
*AppPot* is a way to implement generic application deployment on a computational Grid, and especially to enable users to provide their own software to the computing cluster.

*AppPot* uses [User Mode Linux](http://user-mode-linux.sf.net/) to provide users with a development and running environment they have full control over. User Mode Linux provides a virtualized environment in which users can build and run their own computational applications. AppPot? augments this with scripts that can take a copy of the UserModeLinux? system image and submit it as a regular computational job using the ARC middleware.

An overview of *AppPot* and its purpose can be found in [these slides](http://www.egi.eu/indico/materialDisplay.py?contribId=108&sessionId=47&materialId=slides&confId=452) presented at the [EGI Technical Forum 2011](https://www.egi.eu/indico/conferenceDisplay.py?confId=452).

The most complete information on the AppPot design, motivation, and architecture is the [arXiv 1203.1466 paper](http://arxiv.org/abs/1203.1466).


debian/changelog
apppot (0.27-1) squeeze; urgency=low

  * Put `linux-image-*` packages on hold, so that upgrades 
    from within UML do not try to run `update-grub`. (Thanks 
    to Th. Wuest for the report)
  * `apppot-start` resolves symlinks before starting the VM: changes files
    and job directories should be properly handled even when they are
    symlinks. (Thanks to Th. Wuest for the report)
  * Detect Linux boot partition in disk-based images. (Fixes Issue 10,
    thanks to Tyanko Aleksiev for the fix.)
  * apppot-start: use $APPPOT_MEM to set the default of the `--mem` option.

 -- Riccardo Murri <riccardo.murri@gmail.com>  Thu, 14 Jun 2012 07:58:00 +0200

apppot (0.26-1) squeeze; urgency=low

  * Disable consoles in /etc/inittab on install; re-enable them upon remove.
  * Replace /etc/rc.local if it contains `apppot-init.sh` code.
  * Fix dependencies of the `apppot` package.

 -- Riccardo Murri <riccardo.murri@gmail.com>  Thu, 14 Dec 2011 00:04:00 +0100

apppot (0.25-3) squeeze; urgency=low

  * Fix crash on upgrade.
  * Fix syntax error in init script.

 -- Riccardo Murri <riccardo.murri@gmail.com>  Thu, 8  Dec 2011 14:45:00 +0100

apppot (0.25-0) squeeze; urgency=low

  * Fall back to TERM=vt100 if current $TERM is not supported. Fixes Issue 4.

  * New options `--version` and `--extra` to `apppot-start`.

 -- Riccardo Murri <riccardo.murri@gmail.com>  Thu, 8  Dec 2011 13:05:00 +0100

apppot (0.24-1) squeeze; urgency=low

  * Set architecture to "all", as appropriate for scripts.

 -- Riccardo Murri <riccardo.murri@gmail.com>  Thu, 8  Dec 2011 11:21:35 +0100

apppot (0.24) squeeze; urgency=low

  * Initial release.

 -- Riccardo Murri <riccardo.murri@gmail.com>  Thu, 8  Dec 2011 01:21:35 +0100




apppot-init.sh
#! /bin/sh
#
# Sets up the the AppPot environment, including mounting hostfs
# directories, setting up MPI networking, etc.  and then runs the job
# specified by users on the kernel command-line.  If no job is
# specified, starts SCREEN on the first console.
#
# Author: Riccardo Murri <riccardo.murri@gmail.com>
#
# Copyright (C) 2009-2012 GC3, University of Zurich. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#
VERSION='(SVN $Revision$)'

# PATH should only include /usr/* if it runs after the mountnfs.sh script
PATH=/usr/local/bin:/bin:/usr/bin:/sbin:/usr/sbin
NAME=apppot

echo "==== Starting AppPot $VERSION ..."

# Read configuration variable file if it is present
if [ -r /etc/default/$NAME ]; then
    . /etc/default/$NAME
fi


## defaults

APPPOT_USER=user
APPPOT_GROUP=users
APPPOT_HOME=/home/user


## auxiliary functions

# value EXPR
#
# Given an EXPR of the form `name=value`, output the `value` part.
#
value () {
    echo "$1" | cut -d= -f2-
}


# warn [MSG]
#
# Output MSG to STDERR, with a suitable "warning" prefix.
# If MSG is omitted, read it from STDIN.
#
warn () {
  (echo -n "== WARNING: ";
      if [ $# -gt 0 ]; then echo "$@"; else cat; fi) 1>&2
}


# setup_apppot_user UID GID [USER GROUP]
#
# Create a user with the given UID and GID.  If given, arguments USER
# and GROUP specify the UNIX user name and primary group name; if
# omitted, they default to `user` (user name) and `users` (group
# name).
#
setup_apppot_user () {
    # parse arguments
    uid=${1:?Missing required argument UID to setup_apppot_user}
    gid=${2:?Missing required argument GID to setup_apppot_user}
    user=${3:-user}
    group=${4:-users}

    echo "== Setting up AppPot user $user($uid) in group $group($gid) ..."

    # ensure GROUP exists with the given GID
    if getent group "$group" >/dev/null; then
        # group already exists, ensure given GID
        groupmod --gid "$gid" "$group"
    else
        # create user with given GID
        groupadd --gid "$gid" "$group"
    fi

    # ensure USER exists with the given UID/GID
    if getent passwd "$user" >/dev/null; then
        # user already exists, ensure given UID/GID
        usermod --uid "$uid" --gid "$gid" --shell /bin/sh "$user"
    else
        # create user with given UID/GID
        useradd --create-home --uid "$uid" --gid "$gid" --shell /bin/sh "$user"
    fi

    # ensure USER is authorized to sudo
    fgrep -q "$user" /etc/sudoers \
        || cat >> /etc/sudoers <<__EOF__

# user '$user' enabled by '$0'
$user  ALL=(ALL) NOPASSWD: ALL
__EOF__

    # define $APPPOT_HOME to the user home directory
    export APPPOT_HOME="`getent passwd "$user" | cut -d: -f6`"
    find "$APPPOT_HOME" -xdev -print0 | xargs --null chown "$uid":"$gid"

    # give $user access to the system console
    chown "$user" /dev/console

    # ensure that we can still use `screen` if UID/GID has changed from the default
    rm -rf /var/run/screen/S-$APPPOT_USER /var/run/screen/S-$user

    return 0
}


# mount_hostfs HOSTDIR MOUNTDIR
#
# Mount host directory HOSTDIR onto local mount point MOUNTDIR
#
mount_hostfs () {
    hostdir=${1:?Missing required parameter HOSTDIR to 'mount_hostfs'}
    mountdir=${2:?Missing required parameter MOUNTDIR to 'mount_hostfs'}

    if fgrep -q hostfs /proc/filesystems; then
        echo "== Mounting host directory '$hostdir' on local directory '$mountdir' ..."
        mkdir -p "$mountdir"
        mount -t hostfs -o "$hostdir" host "$mountdir"
    else
        warn "No 'hostfs' support in this kernel, cannot mount '$hostdir' on '$mountdir' ..."
        return 1
    fi
}


# merge_changes FILE
#
# Merge the changes from FILE (path in the host filesystem).
# The named FILE must have been created by 'apppot-snap changes FILE'.
#
merge_changes () {
    changes_file="${1:?Missing required parameter FILE to 'merge_changes'}"

    echo "== Merging changes from host file '$changes_file' ..."
    if mount -t hostfs -o / hostroot /mnt; then
        if [ ! -r "/mnt/$changes_file" ]; then
            warn "Cannot read changes file '$changes_file': not merging changes."
            rc=1
        else
            apppot-snap merge "/mnt/$changes_file"
            rc=$?
        fi
        umount /mnt
        return $rc
    fi
}


# setup_term TERM [DEFAULT]
#
# Set the `TERM` environment variable to the given value,
# or fall back to the DEFAULT if the specified
# terminal is not supported on this system.
#
# If DEFAULT is omitted, use 'vt100'.
#
setup_term () {
    term="${1:?Missing required parameter TERM to 'setup_term'}"
    default_term="${2:-vt100}"

    term_name=`tput -T"$term" longname`
    if [ -z "$term_name" ]; then
        # terminal not supported, fall back to default
        warn "Terminal '$term' not supported, falling back to '$default_term'."
        export TERM="$default_term"
    else
        echo "== Using terminal $name"
        export TERM="$term"
    fi
}


## main

# Set the command line arguments to this shell from
# the ones that were passed to the kernel at boot.
# Return 1 if the `/proc/cmdline` file is not accessible.
#
# In addition, the following parameters are parsed
# and set as environment variables:
#   - `apppot.user`: sets variable APPPOT_USER
#   - `apppot.uid`: sets variable APPPOT_UID
#   - `apppot.group`: sets variable APPPOT_GROUP
#   - `apppot.gid`: sets variable APPPOT_GID
#   - `apppot.jobdir`: sets variable APPPOT_JOBDIR
#   - `apppot.tmpdir`: sets variable APPPOT_TMPDIR
#
if [ -r /proc/cmdline ]; then
    set -- `cat /proc/cmdline`
    while [ $# -gt 0 ]; do
        case "$1" in
            apppot.changes=*) export APPPOT_CHANGESFILE="$(value "$1")" ;;
            apppot.user=*)    export APPPOT_USER="$(value "$1")" ;;
            apppot.uid=*)     export APPPOT_UID="$(value "$1")" ;;
            apppot.group=*)   export APPPOT_GROUP="$(value "$1")" ;;
            apppot.gid=*)     export APPPOT_GID="$(value "$1")" ;;
            apppot.jobdir=*)  export APPPOT_JOBDIR="$(value "$1")" ;;
            apppot.tmpdir=*)  export APPPOT_TMPDIR="$(value "$1")" ;;
            TERM=*)           setup_term "$(value "$1")" ;;
        # `--` means end of kernel params and start of job arguments
            --) shift; break ;;
        esac
        shift
    done
fi

# mount job and scratch directories, if present
[ -n "$APPPOT_JOBDIR" ] && mount_hostfs $APPPOT_JOBDIR $APPPOT_HOME/job
[ -n "$APPPOT_TMPDIR" ] && mount_hostfs $APPPOT_TMPDIR /tmp

# extract a snapshot, if there is any
if [ -n "$APPPOT_CHANGESFILE" ]; then
    merge_changes "$APPPOT_CHANGESFILE"
fi

# ensure that the UID and GID of the user account are the same of the
# mounted "job" directory; we need to do this *after* the changes
# merge because it could have overwritten /etc/passwd
setup_apppot_user \
    ${APPPOT_UID:-1000} ${APPPOT_GID:-1000} \
    ${APPPOT_USER:-user} ${APPPOT_GROUP:-users}

# now set arguments according to the kernel command-line (and remove
# quotes that apppot-start.sh put there)
eval set -- "$@"

# run a job or start an interactive shell
if [ $# -eq 0 ]; then
    if [ -x "$APPPOT_HOME/job/apppot-run" ]; then
        # run the job script as the specified user
        echo "== Running job script '$APPPOT_HOME/job/apppot-run' ..."
        su -l "$APPPOT_USER" -c "$APPPOT_HOME/job/apppot-run"
    elif [ -x "$APPPOT_HOME/apppot-autorun" ]; then
        # run the autostart script as the specified user
        echo "== Running autostart script '$APPPOT_HOME/apppot-autorun' ..."
        su -l "$APPPOT_USER" -c "$APPPOT_HOME/apppot-autorun"
    else
        echo "== Starting shell on /dev/console ..."
        su -l "$APPPOT_USER" -c '/usr/bin/screen /bin/bash'
    fi
else
    # execute command-line
    echo "== Running command-line \"$@\" ..."
    if mountpoint -q "$APPPOT_HOME/job"; then
        # if /home/user/job is mounted, execute job in it
        eval su -l "$APPPOT_USER" -c "'cd $APPPOT_HOME/job; $@'"
    else
        # otherwise, execute in $HOME directory
        eval su -l "$APPPOT_USER" -c "'cd $APPPOT_HOME; $@'"
    fi
fi

echo "==== AppPot done, commencing shutdown ..."
# sleep a few seconds to give /sbin/init time to end the boot sequence
# before shutting down the system
(sleep 1; halt) &
return 0


trunk/apppot-snap.sh
#! /bin/sh
#
# Snapshotting support for AppPot VMs.
#
# Author: Riccardo Murri <riccardo.murri@gmail.com>
#
#
# Copyright (C) 2009-2011 GC3, University of Zurich. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#
PROG="$(basename $0)"
VERSION='(SVN $Revision$)'

usage () {
cat <<EOF
Usage:
  $PROG base
  $PROG changes [FILE]
  $PROG merge [FILE]

The function performed by this script depends on the action word that
begins its invocation:

  base      Record the current system state; subsequent 'changes'
            invocations produce an archive file with the differences
            from the last recorded base state.

  changes   Make an archive file at FILE containing the differences from
            the last snapshot created with '$PROG base'.  
            If FILE is omitted, defaults to '$HOME/job/apppot.YYYY-MM-DD.changes.tar.gz'
            (if '$HOME/job' is a mountpoint for accessing the host filesystem)
            or, failing that, a file named 'apppot.changes.tar.gz' in the current
            directory.

  merge     Merge differences from a the archive file at FILE
            into the current system.
            If FILE is omitted, defaults to '$HOME/job/apppot.YYYY-MM-DD.changes.tar.gz'
            (if '$HOME/job' is a mountpoint for accessing the host filesystem)
            or, failing that, a file named 'apppot.changes.tar.gz' in the current
            directory.


Options:

  -v, --verbose     Verbosely print what files are being saved/restored.

  -n, --just-print  Do not make any modification to the system;
                    just print what would have been done.

  -h, --help        Print this help text.

EOF
}


## defaults

base_state_file='/var/lib/apppot/apppot-snap.base'
exclude_file='/var/lib/apppot/apppot-snap.exclude'


## helper functions
die () {
  rc="$1"
  shift
  (echo -n "$PROG: ERROR: ";
      if [ $# -gt 0 ]; then echo "$@"; else cat; fi) 1>&2
  exit $rc
}

warn () {
  (echo -n "$PROG: WARNING: ";
      if [ $# -gt 0 ]; then echo "$@"; else cat; fi) 1>&2
}

have_command () {
  type "$1" >/dev/null 2>/dev/null
}

require_command () {
  if ! have_command "$1"; then
    die 1 "Could not find required command '$1' in system PATH. Aborting."
  fi
}

is_absolute_path () {
    expr match "$1" '/' >/dev/null 2>/dev/null
}

home_directory () {
    # if we're running under `sudo`, we want to use the real user's
    # $HOME, not `root`'s one...
    if [ -n "$SUDO_USER" ]; then
        echo $(getent passwd "$SUDO_USER" | cut -d: -f6)
    else
        echo $HOME
    fi
}

## parse command-line 

maybe=''
verbose=''

if [ "x$(getopt -T)" = 'x--' ]; then
    # old-style getopt, use compatibility syntax
    set -- $(getopt 'hnv' "$@")
else
    # GNU getopt
    eval set -- $(getopt --shell sh -l 'just-print,help,no-act,verbose' -o 'hnv' -- "$@")
fi
while [ $# -gt 0 ]; do
    case "$1" in
        --just-print|--no-act|-n) maybe='echo' ;;
        --help|-h) usage; exit 0 ;;
        --verbose|-v) verbose='-v' ;;
        --) shift; break ;;
    esac
    shift
done

action=$1

## main

require_command date
require_command tar
require_command mktemp
require_command mountpoint

if ! [ -d $(dirname "$base_state_file") ]; then
    die 1 "Directory '$(dirname "$base_state_file")' that should host snapshot files, is non-existent."
fi

if ! [ -r "$exclude_file" ]; then
    die 1 "Exclude file '$exclude_file' cannot be read."
fi

default_archive () {
# let the default archive location be what the init script expects
    home="$(home_directory)"
    if mountpoint -q "$home/job"; then
        echo "$home/job/apppot.$(date -I).changes.tar.gz"
    else
        echo "$(pwd)/apppot.changes.tar.gz"
        warn "Default archive location '$home/job' does not look like a hostfs mount, storing changes into '$default_archive' instead."
    fi
}

case $action in

    base) 
        if test -r "$base_state_file"; then
            warn "Deleting old base state file '$base_state_file', interrupting this command may leave the system without a valid base state."
            rm -f "$base_state_file"
        fi
        tar $verbose --create -f /dev/null -C / . \
            --listed-incremental="$base_state_file" \
            --exclude-from="$exclude_file" --exclude-backups \
            --anchored --wildcards-match-slash
        ;;

    changes) 
        archive="${2:-$(default_archive)}"
        # check that the snapfile exists and can be read
        test -r "$base_state_file" \
            || die 1 "Base state file '$base_state_file' does not exist."
        # create a temporary copy of the tar snapfile, as tar modifies
        # it during a `--create` operation
        snapfile=`mktemp` \
            || die 1 "Cannot create temporary state file."
        trap "{ rm -f '$snapfile'; }" EXIT
        $maybe cp $verbose "$base_state_file" "$snapfile" \
            || die 1 "Cannot copy base state file to temporary location '$snapfile'."
        $maybe tar $verbose --create --auto-compress -f "$archive" -C / . \
            --listed-incremental="$snapfile" \
            --one-file-system \
            --exclude-from="$exclude_file" --exclude-backups \
            --exclude="$snapfile" --exclude="$archive" --exclude="$exclude_file" \
            --anchored --wildcards-match-slash
        echo "Saved changes from base state into file '$archive'."
        ;;

    merge) 
        archive="${2:-$(default_archive)}"
        $maybe tar $verbose --extract --auto-compress -f "$archive" -C / \
            --same-owner --same-permissions \
            --listed-incremental="$base_state_file" \
            --exclude-from="$exclude_file" --exclude-backups \
            --exclude="$snapfile" --exclude="$archive" --exclude="$exclude_file" \
            --anchored --wildcards-match-slash
        ;;

    *)
        die 1 "Unknown action '$action'; type '$PROG --help' for usage help."
        ;;

esac

apppot-start.sh
#! /bin/sh
#
# Start an AppPot system image to execute a given program with
# specified arguments.
#
# Author: Riccardo Murri <riccardo.murri@gmail.com>
#
#
# Copyright (C) 2009-2012 GC3, University of Zurich. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#
PROG="$(basename $0)"
VERSION='(SVN $Revision$)'

usage () {
cat <<EOF
Usage: $PROG [options] [PROG ARGS...]

Start an AppPot system image to execute PROG with the given ARGS.
If PROG and ARGS are omitted, then start the AppPot system image
with an interactive console.

If a valid MPI host file is found, connect interface 'eth1' in the
AppPot system image to the local multicast address 239.255.82.77 on
port 8277.

The following options are recognized; option processing stops at the
first non-option argument (which must be PROG):

  --apppot PATH   Use the specified AppPot system image.
                  You can also specify a pair COWFILE:IMAGEFILE,
                  in which case IMAGEFILE will be opened read-only
                  and all changes will be written to COWFILE instead.

  --changes FILE  Merge the specified changes file into the AppPot system
                  image.  FILE must have been created with the
                  'apppot-snap changes' command (which see).

  --mem NUM       Amount of memory to allocate to the
                  AppPot system image; use the 'M' or 'G'
                  suffix to denote MB or GB respectively.

  --slirp PATH    Use the executable found at PATH as
                  the 'slirp' command for providing
                  IP network access.

  --uml PATH      Use the UML 'linux' executable found at PATH.

  --id NAME       Use NAME to control the running instance
                  with 'uml_mconsole'.

  --extra ARG     Append ARG to the UML kernel command-line.
                  (Repeat option multiple times to append more ARGs)

  --version, -V   Print version and exit.

  --help, -h      Print this help text.

EOF
}


## helper functions
die () {
  rc="$1"
  shift
  (echo -n "$PROG: ERROR: ";
      if [ $# -gt 0 ]; then echo "$@"; else cat; fi) 1>&2
  exit $rc
}

warn () {
  (echo -n "$PROG: WARNING: ";
      if [ $# -gt 0 ]; then echo "$@"; else cat; fi) 1>&2
}

have_command () {
  type "$1" >/dev/null 2>/dev/null
}

link_target () {
  find "$1" -printf '%l'
}

quote () {
    echo "$1" | sed -e 's|\\|\\\\|g;s|"|\\"|g;'
}

require_command () {
  if ! have_command "$1"; then
    die 1 "Could not find required command '$1' in system PATH. Aborting."
  fi
}

stdin_is_not_dev_null () {
  test "$(link_target /proc/self/fd/0)" != '/dev/null'
}

is_absolute_path () {
    expr match "$1" '/' >/dev/null 2>/dev/null
}


## parse command-line

# defaults
if [ -n "$APPPOT_IMAGE" ]; then
    apppot="$APPPOT_IMAGE"
else
    apppot="apppot.img"
fi

if [ -n "$APPPOT_MEM" ]; then
    mem="$APPPOT_MEM"
else
    mem="512M"
fi

extra=''

# try to provide sensible defaults
if [ -x "`pwd`/slirp" ]; then
    slirp="`pwd`/slirp"
elif have_command slirp-fullbolt; then
    slirp='slirp-fullbolt'
elif have_command slirp; then
    slirp='slirp'
else
    slirp=''
fi

if [ -x "`pwd`/linux" ]; then
    linux="`pwd`/linux"
elif [ -n "$APPPOT_KERNEL" ]; then
    linux="$APPPOT_KERNEL"
elif have_command linux; then
    linux="linux"
else
    linux=''
fi

# parse command line
while [ $# -gt 0 ]; do
    case "$1" in
        --apppot|--image) shift; apppot="$1" ;;
        --changes|--merge) shift; changes="$1" ;;
        --extra|-X) shift; extra="$extra $1" ;;
        --id|--umid) shift; umid="$1" ;;
        --mem) shift; mem="$1" ;;
        --slirp) shift; slirp="$1" ;;
        --linux|--kernel|--uml) shift; linux="$1" ;;
        --help|-h) usage; exit 0 ;;
        --version|-V) echo "$PROG $VERSION"; exit 0 ;;
        --*) die 1 "Unknown option '$1'; type '$PROG --help' to see usage help." ;;
        --) shift; break ;;
        # parsing stops at the first non-option argument
        *) break ;;
    esac
    shift
done


## main

require_command expr
require_command file
require_command hostname
require_command readlink

unique="$(hostname).$$"

# parse the `--apppot` argument and check for existence of the backing file
case "$apppot" in
    # UMLx allows both `ubdX=cow,back` and `ubdX=cow:back`
    *,*)
        apppot_cow="$(echo "$apppot" | cut -d, -f1)"
        apppot_img="$(echo "$apppot" | cut -d, -f2)"
        ;;
    *:*)
        apppot_cow="$(echo "$apppot" | cut -d: -f1)"
        apppot_img="$(echo "$apppot" | cut -d: -f2)"
        ;;
    # else there's a single file, which is opened R/W
    *)
        apppot_cow=''
        apppot_img="$apppot"
        ;;
esac
if ! [ -r "$apppot_img" ]; then
    die 1 "Cannot read AppPot image file '$apppot_img' - aborting."
fi
if [ -z "$apppot_cow" ] && [ ! -w "$apppot_img" ]; then
    apppot_cow="apppot.$(hostname).$$.cow"
    warn "AppPot image file '$apppot_img' is read-only: writing changes to COW file '$appot_cow'."
    apppot="$apppot_cow:$apppot_img"
fi


if [ -z "$linux" ]; then
    die 1 "No 'linux' executable detected, please specify the UML kernel via the '--linux' option."
else
    require_command $linux
fi

if [ -z "$slirp" ]; then
    warn "No 'slirp' or 'slirp-fullbolt' executable detected, disabling network access."
else
    require_command $slirp
    opt_slirp="eth0=slirp,,$slirp"
fi

if [ -z "$umid" ]; then
    umid="apppot.${unique}"
fi

if [ -n "$changes" ]; then
    # ensure `apppot.changes` points to an absolute path
    _changes="$(readlink --canonicalize-existing ${changes})"
    if [ $? -ne 0 ]; then
        die 1 "Cannot resolve path to changes file '${changes}' (see above)."
    fi
    opt_changes="apppot.changes=${_changes}"
fi

# determine whether $apppot is a filesystem or disk image,
# and generate a `root=...` kernel parameter
# XXX: complicated heuristics, may fail unpredictably
what=$(file --dereference --brief "$apppot_img")
case "$what" in
    'x86 boot sector'*)
        # disk image, determine boot partition
        bootpart=$(/usr/bin/expr match "$what" '.*partition \([1-4]\): ID=0x83, active')
        if [ -n "$bootpart" ]; then
            rootfs="/dev/ubda$bootpart"
        else
            warn "Disk image '$apppot_img' contains no bootable partition, assuming Linux is on partition 1."
            rootfs="/dev/ubda1"
        fi
        ;;
    *'filesystem data'*)
        # filesystem image
        rootfs=/dev/ubda
        ;;
esac


# gather environmental information
require_command id
APPPOT_UID=`id -u`
APPPOT_GID=`id -g`

if [ -n "$TERM" ]; then
    opt_term="TERM=$TERM"
fi

jobdir="$(readlink --canonicalize .)"


# prepare command-line invocation
cmdline=''
for arg in "$@"; do
    cmdline="$cmdline '$(quote $arg)'"
done

# UMLx cannot use stdin for console input if it is connected to a file
# or other no-wait stream (e.g., /dev/null); in order to make sure
# that this startup script can run with STDIN connected to any stream,
if test -t 0; then
    # STDIN is a terminal, start UMLx as usual
    $linux \
        umid="$umid" \
        highres=off \
        mem="$mem" \
        hostfs=/ \
        ubd0="$apppot" \
        "$opt_slirp" \
        eth1=mcast,,239.255.82.77,8277,1 \
        con=fd:0,fd:1 \
        root="$rootfs" \
        "$opt_term" \
        "$opt_changes" \
        $extra \
        apppot.uid=$APPPOT_UID \
        apppot.gid=$APPPOT_GID \
        apppot.jobdir="${jobdir}" \
        -- "$cmdline"

# STDIN is not a terminal; what we can do now depends on the
# availability of the `empty` helper command.
#
# See http://empty.sourceforge.net/ or install Debian/Ubuntu package
# 'empty-expect'.
#

elif have_command empty && stdin_is_not_dev_null; then
    # save STDIN for later use with `empty -s`
    exec 3<&0

    # detach from the input stream
    exec < /dev/null

    # start UMLx
    empty -f -i ".apppot.${unique}.stdin" -o ".apppot.${unique}.stdout" $linux \
        umid="$umid" \
        mem="$mem" \
        hostfs=/ \
        ubd0="$apppot" \
        "$opt_slirp" \
        eth1=mcast,,239.255.82.77,8277,1 \
        con=fd:0,fd:1 \
        root="$rootfs" \
        "$opt_term" \
        "$opt_changes" \
        $extra \
        apppot.uid=$APPPOT_UID \
        apppot.gid=$APPPOT_GID \
        apppot.jobdir="${jobdir}" \
        -- "$cmdline"

    # send original STDIN to the UMLx through the named pipe
    (empty -s -o ".apppot.${unique}.stdin" -c 0<&3) &

    # send UMLx console output to STDOUT
    cat ".apppot.${unique}.stdout"


else
    require_command kill
    require_command mkfifo
    require_command sleep

    # No helper programs, we use the following trick:
    #   - create a named FIFO
    #   - connect a process that writes no output to the write end of the FIFO
    #   - connect the UMLx instance to the read end of the FIFO,
    #     and use STDIN/STDOUT for the system console
    #
    mkfifo ".apppot.${unique}.stdin" \
        || die 1 "Cannot create FIFO '`pwd`/.apppot.${unique}.stdin'"
    (sleep 365d) > ".apppot.${unique}.stdin" &
    stdin_pid=$!

    # ensure the FIFO is removed and the `sleep` process is killed
    cleanup () {
        kill $stdin_pid
        rm -f ".apppot.${unique}.stdin"
    }
    trap "cleanup" EXIT

    # I found no way of conveying arbitrary STDIN content into the
    # named FIFO; so this trick only works for simulating a null
    # STDIN, so let's warn users.
    #
    if stdin_is_not_dev_null; then
        warn "Redirecting output from /dev/null, any content to STDIN will be lost."
    fi

    # start UMLx with input from the FIFO
    $linux \
        umid="$umid" \
        mem="$mem" \
        hostfs=/ \
        ubd0="$apppot" \
        "$opt_slirp" \
        eth1=mcast,,239.255.82.77,8277,1 \
        con=fd:0,fd:1 \
        root="$rootfs" \
        "$opt_term" \
        "$opt_changes" \
        $extra \
        apppot.uid=$APPPOT_UID \
        apppot.gid=$APPPOT_GID \
        apppot.jobdir="${jobdir}" \
        -- "$cmdline" \
        < ".apppot.${unique}.stdin"
fi
